### PR TITLE
feat(replication): add replica/ack count to range

### PR DESCRIPTION
**Pull request by superhx**

### CODE REVIEW

File: components/replication/src/stream_manager/replication_range.rs [hunk 1]
```diff
@@ -7,31 +7,36 @@ use std::{
...
 pub(crate) struct ReplicationRange {
-    metadata: Range,
+    metadata: RangeMeta,
 
     client: Weak<Client>,
 
     replicators: Vec<Replicator>,
 
-    /// Inflight append entry requests.
+    /// Inflight append entry requestscomponents/client/src/client/client.rs.
```
Comment (Chillax-0v0) [inline components/replication/src/stream_manager/replication_range.rs:24]: "components/client/src/client/client.rs"

File: components/model/src/range/mod.rs [hunk 2]
```diff
@@ -29,9 +30,18 @@ pub struct Range {
...
+    /// The range replica expected count. When cluster nodes count is less than replica count but
+    /// but larger than ack_count, the range can still be successfully created.
+    replica_count: u32,
+
+    /// The range replica ack count, only success ack > ack_count, then the write is success.
+    /// For seal range, success seal range must seal replica count >= (replica_count - ack_count + 1)
+    ack_count: u32,
 }
 
-impl Range {
```
Comment (lizhanhui) [inline components/model/src/range/mod.rs:34]: 如果表示Range元数据的话，使用完整拼写？ RangeMeta --> RangeMetadata

File: components/model/src/range/mod.rs [hunk 3]
```diff
@@ -40,6 +50,29 @@ impl Range {
             start,
             end,
             replica: vec![],
+            replica_count: 0,
+            ack_count: 0,
+        }
+    }
+
+    pub fn new_range(
```
Comment (lizhanhui) [inline components/model/src/range/mod.rs:58]: Rust里面的通用规则，一般默认的构造叫new. 其它定制化的构造器，with_XXX

Comment (lizhanhui) [reply components/model/src/range/mod.rs:58]: 比如： https://docs.rs/bytes/latest/bytes/struct.BytesMut.html#method.with_capacity

File: components/replication/src/stream_manager/replication_range.rs [hunk 4]
```diff
@@ -52,16 +57,27 @@ impl ReplicationRange {
         Ok(())
     }
 
-    fn confirm_offset(&self) -> Result<u64, ReplicationError> {
+    fn cal_confirm_offset(&self) -> Result<u64, ReplicationError> {
```
Comment (lizhanhui) [inline components/replication/src/stream_manager/replication_range.rs:60]: 名字咱们约定拼写全吧，后期好维护一些

File: components/model/src/range/mod.rs (hunk 3 above)
Comment (superhx) [reply components/model/src/range/mod.rs:58]: 待新增 RangeReplicaMetaData 后，new_range 会替换成 new，暂时保持现状

File: components/model/src/range/mod.rs (hunk 2 above)
Comment (superhx) [reply components/model/src/range/mod.rs:34]: 已修改

File: components/replication/src/stream_manager/replication_range.rs (hunk 4 above)
Comment (superhx) [reply components/replication/src/stream_manager/replication_range.rs]: 已修改

File: components/protocol/fbs/rpc.fbs [hunk 5]
```diff
@@ -297,6 +297,12 @@ table Range {
 
     // The collection of data-node that host this range.
     nodes: [DataNode] (id: 5);
+
+    // The range replica expected count.
+    replica_count: int32 = -1 (id: 6);
```
[the remark this answers is not in the document]
Comment (superhx) [reply components/protocol/fbs/rpc.fbs:302]: When cluster nodes is less than replica_count, but larger than ack_count, the nodes.len() != replica_count

File: components/protocol/fbs/rpc.fbs [hunk 6]
```diff
@@ -297,6 +297,12 @@ table Range {
...
+
+    // The range replica ack count, only success ack > ack, then the write is success.
+    ack_count: int32 = -1 (id: 7);
```
[the remark this answers is not in the document]
Comment (superhx) [reply components/protocol/fbs/rpc.fbs:305]: When list range, SDK need ack_count to seal the last unsealed range

File: components/replication/src/stream_manager/replication_range.rs [hunk 7]
```diff
@@ -52,16 +57,27 @@ impl ReplicationRange {
...
             return Err(ReplicationError::Internal);
         }
 
+        // Example1: replicas confirmOffset = [1, 2, 3]
+        // - when replica_count=3 and ack_count = 1, then result confirm offset = 3.
+        // - when replica_count=3 and ack_count = 2, then result confirm offset = 2.
+        // - when replica_count=3 and ack_count = 3, then result confirm offset = 1.
+        // Example2: replicas confirmOffset = [1, corrupted, 3]
+        // - when replica_count=3 and ack_count = 1, then result confirm offset = 3.
+        // - when replica_count=3 and ack_count = 2, then result confirm offset = 2.
```
Comment (Chillax-0v0) [inline components/replication/src/stream_manager/replication_range.rs:71]: When
confirmOffset = [1, corrupted, 3], 
replica_count = 3,
ack_count = 2,

result confirm offset should be **3**

File: components/replication/src/stream_manager/replication_range.rs [hunk 8]
```diff
@@ -52,16 +57,27 @@ impl ReplicationRange {
...
         self.replicators
             .iter()
+            .filter(|r| !r.corrupted())
             .map(|r| r.confirm_offset())
-            .sorted() // Ascending order
-            .nth((self.replicators.len() - 1) / 2)
+            .sorted()
+            .rev() // Descending order
+            .nth(confirm_offset_index as usize)
             .ok_or(ReplicationError::Internal)
```
Comment (Chillax-0v0) [inline components/replication/src/stream_manager/replication_range.rs:81]: When
replicas confirmOffset = [1, corrupted, 3],
replica_count = 3,
ack_count = 1,

this func returns ReplicationError. However, it should return 3.

Maybe we can add some tests on this func?

File: components/replication/src/stream_manager/replication_range.rs (hunk 7 above)
Comment (superhx) [reply components/replication/src/stream_manager/replication_range.rs]: the result confirm offset should be 1

Comment (superhx) [reply components/replication/src/stream_manager/replication_range.rs]: the result confirm offset should be 1

File: components/replication/src/stream_manager/replication_range.rs (hunk 8 above)
Comment (superhx) [reply components/replication/src/stream_manager/replication_range.rs]: fixed with confirm_offset_index = ack_count - 1. (Add unit test in next PR)

**Review by Chillax-0v0 — APPROVED** [review]

**Pull request merged**